cosigner/authstatestore.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2024 OpenPubkey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package cosigner
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type AuthStateStore interface {
+	CreateNewAuthSession(pkt *pktoken.PKToken, ruri string, nonce string) (authID string, err error)
+	LookupAuthState(authID string) (*AuthState, bool)
+	UpdateAuthState(authID string, authState AuthState) error
+	CreateAuthcode(authID string) (authcode string, err error)
+	RedeemAuthcode(authcode string) (authState AuthState, authID string, err error)
+}
+
+// This is intended for testing purposes. The locking strategy used is not
+// particularly efficient. Anyone building a Cosigner should use the interface
+// above to replace this in-memory store with a database.
```

Comment:
I think my one big ask is to move any code that this comment applies to into the examples folder. By its nature, this repo will benefit from a very clear demarcation between library code and implementations. I think there's maybe just two things that warrant the move:

  1. The implementation of `AuthStateInMemoryStore`. That would leave this file with just the interface definition, which then might be best moved to authcosigner.go
  2. Possibly the specific implementation of `AuthIDIssuer`?

But once you do that, you may conclude that some other things also count as example code.

Reply:
Moved `AuthStateInMemoryStore` to `cosigner/mocks` as it was originally intended to for testing and then repurposed by example MFA Cosigner. Let me know if that resolves this issue for you.

`AuthIDIssuer` is intended for production use. I wanted an atomic increment unique ID generator that is 256-bit in length and I didn't want to assume everyone will use a SQL database backend for this purpose.